### PR TITLE
[PyTorch] Add test_modules test for TransformerEncoderLayer fast path

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78269
* __->__ #78268

Extend the existing TransformerEncoderLayer test to cover the fast path.

Differential Revision: [D36564009](https://our.internmc.facebook.com/intern/diff/D36564009/)

### CODE REVIEW

File: torch/testing/_internal/common_modules.py [hunk 1]
```diff
@@ -678,6 +678,29 @@ def module_inputs_torch_nn_TransformerEncoderLayer(module_info, device, dtype, r
                 desc='no_batch_dim'
             ))
 
+    # test fast path. We abuse the fact that we can supply an arbitrary reference_fn to
+    # force the fast path to run -- we cannot call .eval() through the ordinary
+    # ModuleInfo interface, so the "reference" is actually the code under test.
```
Comment (jbschlosser) [inline torch/testing/_internal/common_modules.py:683]: We should just fix the `ModuleInfo` interface so it's flexible enough to do what you need here.

Maybe an enum on `ModuleInput` specifying whether it should be used for training / evaluation / both (default) ? If this works okay, I'll open a PR for the support.

Comment (jbschlosser) [reply torch/testing/_internal/common_modules.py:683]: Addressing this in #78735.

Comment (jbschlosser) [reply torch/testing/_internal/common_modules.py:683]: @swolchok FYI this landed. Tests now run in both train and eval, and you can add eval-only sample inputs

**Review by zrphercule — APPROVED** [review]
LGTM, Thanks!

**Pull request closed without merging**